### ISSUE TITLE
Improve method for adding colour to table cells

### ISSUE THREAD

**Issue by rockyallen** [asciidoctor/asciidoctor#1338]
Currently the syntax for doing this is described in the Wiki at https://github.com/asciidoctor/asciidoctor/wiki/How-to-set-the-background-color-of-a-table-cell.
It is a bit clunky, and limited in scope (see http://discuss.asciidoctor.org/Using-cellbgcolor-to-colour-table-cells-that-have-asciidoc-content-tt3143.html) .

An improved implementation should:
- Apply to all backends that are capable of displaying colour (eg DocBook/HTML, DocBook/pdf, HTML and PDF);
- Work with any type of cell content (literal or Asciidoc);
- Allow the foreground colour to be changed?

If the syntax is to be changed (Dan's preference is to add a "role" attribute to the cells and process it using CSS), it should ideally:
- Apply at table, column, and cell level, with the narrower scope overriding the wider;
- Be conceptually similar to the way the other formatting (eg alignment, width) is applied;
- Not require changes to DocBook customisation layers.

Any thoughts? Especially additional use cases and requirements.

My tuppence worth: I quite like the current method, as it keeps the formatting attributes independent and contained entirely within the Asciidoc. If roles were used, would I then have to modify all my DocBook customisation layers to process them?

**Comment by mojavelinux** [comment]
I was thinking about enhancing the cell formatting instructions by allowing a role to be applied to a cell. Then, you can use CSS to set the color as well as any other formatting you want.

We just need to think about where the role would fit. Let's look at a cell that has all the possible settings applied:

```
2.3+<.>m| content
```

We need a place for the role to go. Perhaps we can put it after the style parameter.

```
2.3+<.>m.result| content
```

Then, for a cell without any other formatting, it would look like:

```
.result| content
```

That seems pretty reasonable. However, it does run the risk of matching where it shouldn't (other cell content).

We could enforce that it must be directly adjacent to the pipe (which I think is the case already). That reduces the chance for error.

Another option is to add square brackets around the role like for an inline phrase.

```
[result]| content
```

In the extreme case, that would look like:

```
2.3+<.>m[result]| content
```

We may have to try both options to see which one creates the least amount of conflict.

**Comment by wolandscat** [comment]
I'm not too worried about the syntax being arcane - custom-shaded tables are arcane. But we need them. I'm most interested in there being a way to do a table like [this one](https://rawgit.com/openEHR/specifications-RM/master/docs/ehr/ehr.html#_ehr_status_class) - note some rows have shading right across, some just in column 1.

We generate these tables using a UML tool extractor, so arcane syntax doesn't matter - just that we can get the result.

**Comment by mojavelinux** [comment]
>  But we need them.

I agree. We need roles on table cells for the same reason we have roles on blocks and such.

> I'm most interested in there being a way to do a table like this one - note some rows have shading right across, some just in column 1.

I still hold to my opinion that you already have what you need to accomplish this particular use case. These are header cells. Add the header (i.e., "h") style to them and style the `tbody th` element in the stylesheet. If you need more fine-grained control, add a role to the table so you know which tables should receive this treatment.

If you need help implementing my recommendation, please let me know. I'd be glad to help clarify.

**Comment by wolandscat** [comment]
Oops - yes you did say that somewhere and I didn't absorb it. You meant 'header' as in 'functioning as a header kind of cell' not literally 'top of the table' cells. I'll raise a CR on our project to get this implemented on the generator.

**Comment by Svish** [comment]
Re #1564, please see if you can make roles supported for table rows as well, not just single cells?

**Comment by mojavelinux** [comment]
The idea for supporting attributes on a row is to put it above the line of the first cell, like so:

```
[.row-role]
| content
```

The role for cell would have to be on the same line as the pipe.

I do want to mention that this is going to be very difficult to implement in the current processor because it's going to impose a more rigid structure than what is currently there. This is definitely something we want to do for UniDoc where we can make more drastic changes like that. I'll try to get it into Asciidoctor before then, but it might be more limited.

**Comment by mojavelinux** [comment]
In other words, in UniDoc, we want tables to have a much more well-defined structure. Currently, you can do crazy things like put cell delimiters anywhere on the line and with all kinds of space and wrapping permitted.

**Comment by JBR69** [comment]
I wrote this to use specific role in a cell (asciidoctor-1.5.5)

in **asciidoctor.rb** line 775, replace 
```
CellSpecStartRx = /^#{CG_BLANK}*(?:(\d+(?:\.\d*)?|(?:\d*\.)?\d+)([*+]))?([<^>](?:\.[<^>]?)?|(?:[<^>]?\.)?[<^>])?([a-z])?$/
```
with
```
CellSpecStartRx = /^#{CG_BLANK}*(?:(\d+(?:\.\d*)?|(?:\d*\.)?\d+)([*+]))?([<^>](?:\.[<^>]?)?|(?:[<^>]?\.)?[<^>])?([a-z])?(\[.*\])?$/
```
--> add `(\[.*\])?` before `$/`

in **table.rb** line 230, add
```
# Public: Get/Set the role for this cell (default: nil)
attr_accessor :role
```

in **table.rb** line 237, add
```
@role = nil
```

in **table.rb** line 250, add
```
@role = attributes['role'] if attributes.key? 'role'
```

in **parser.rb** line 2518, add
```
if m[5]
	spec['role'] = m[5].gsub("[", "").gsub("]", "")
end
```

in **html.rb** line 827, replace 
```
cell_class_attribute = %( class="tableblock halign-#{cell.attr 'halign'} valign-#{cell.attr 'valign'}")
```
with
```
cell_class_attribute = %( class="tableblock halign-#{cell.attr 'halign'} valign-#{cell.attr 'valign'} #{cell.attr 'role'}")
```
--> add `#{cell.attr 'role'}` before `")`

-----

In my adoc
```
[cols="50%, 50%",]
|==============================
| Text1-1
[greenCell]| Text1-2

^[greenCell]| Text2-1
^| Text2-2
|==============================
```

and in my CSS
```
.greenCell {background-color: #779912;}
```

constraint : the role must be between `[` and `]` and just before the `|`

Is this your approach ?

Jérôme

**Comment by brathax** [comment]
Another thing that is not possible at the moment is to influence the border of cells or rows. In Docbook you can apply rowsep und colsep on `<table>`, `<row>` and `<entry>`. In HTML a special class could be emitted to make styling via CSS border properties possible.

This would enable styling possibilities with tables that look like this
```
===========
Head1 Head2
-----------
content
content
content
===========
```

or multi-row table headers

```
===========
  Category
Head1 Head2
-----------
content
content
content
===========
```

**Comment by wolandscat** [comment]
Any progress on this issue? I'm wondering what method can be used to control different tables to use different header background colours. I think that would still require a role somewhere in a table coming after the 'h' or else associated with options="header".

**Comment by Tobski** [comment]
This would be an *extremely* useful feature to have - we're doing automatic processing of asciidoc content and a couple of the tools we have are kind of broken without this (and it's blocking from doing a few new bits of tooling as well). Is this likely to be on the radar any time soon?

**Comment by PauloFrancaLacerda** [comment]
I was looking for an open topic for this subject; almost opened a brand new one. Luckly, I found this one.

Currently, I'm using an (Windows-only) ad-hoc HTML post-processor to convert my own "directives" into cell role specifiers. It works fine for my needs, but the adoc code gets quite messy.

It would be very helpful to have this feature (cell role specifiers) as a built-in feature.

**Comment by elextr** [comment]
When Unidoc is specified, if it uses @mojavelinux suggestion of the square bracketed role, that could be expanded to be a full attribute list so other rare stuff could be added without more syntax breaking.

**Comment by brunchboy** [comment]
I think that once this is implemented, along with #1564, I will have the pieces I need to port my LaTeX [bytefield](http://mirrors.ibiblio.org/CTAN/macros/latex/contrib/bytefield/bytefield.pdf) diagrams to Asciidoctor, and then can convert my two big LaTeX articles to Asciidoctor documents, and serve them in convenient HTML rather than clunky PDF (and have one less tool chain and document editing syntax for collaborators to install and learn). So I have my fingers crossed that this can be done!

**Comment by brunchboy** [comment]
> Any progress on this issue? I'm wondering what method can be used to control different tables to use different header background colours. I think that would still require a role somewhere in a table coming after the 'h' or else associated with options="header".

@wolandscat your goals can be accomplished today by using CSS: I was just doing something similar by assigning a role to the table, which is currently possible, and having CSS rules that override the standard `.doc table.tableblock th` rules. My `.doc table.bytefield th` rules would win whenever I assigned a table the role `bytefield`. I only got stuck when I needed individual cells to have roles to override their styles, which is what brought me here.

**Comment by fiddlededee** [comment]
Hi. We also need this functionality very much (we use Asciidoctor as a report generator via direct html and pdf-fopub toolchains, so straightforward way to format cells is essential for us), but we've got no Ruby devs in our organization. Could someone suggest a person or organization that is able to hack  Asciidoctor the right way to implement the feature. As I've understood, @mojavelinux says that it is difficult to implement this on the level of rows, but the level of cells would suffice. 

For now we are using something like @brunchboy suggestion. But on the level of Docbook h| turns not into a role, but into an emphasized text inside the cell. So we had to make docbook-xsl ignore standard way to emphasize text in such cells and introduce our own role for bold font. Clearly doesn't look like satisfactory solution. Once again h| -- is only one role. But very often we need more than one.

**Comment by Mogztter** [comment]
> Could someone suggest a person or organization that is able (on freelance terms) to hack Asciidoctor the right way to implement the feature. 

I guess you should contact https://opendevise.com/ the company behind Asciidoctor.

**Comment by Mogztter** [comment]
> Apply at table, column, and cell level, with the narrower scope overriding the wider;

Do we really want that?
I think it can be useful to append roles (especially if you want to apply a common style using CSS). Let's take an example:

```
[cols="1,1[amount]"]
|===
[starfighter]|Firespray-class interceptor
|25 Republic credits
        
[bomber]|BTL-A4 Y-wing 
|100 Republic credits

[total]|Total
[sum]|125 Republic credits
|===
```

In the example above, I've defined the following roles:

- `amount` on the second column,  
- `starfighter` on the first column of the first row
- `bomber` on the first column of the second row
- `total` on the first column of the third row 
- `sum` on the second column of the third row

In my opinion, the second column on the third row should have both the `amount` and the `sum` role.

**Comment by mojavelinux** [comment]
> Could someone suggest a person or organization that is able (on freelance terms) to hack Asciidoctor the right way to implement the feature.

This will be considered as part of the AsciiDoc spec. It will not move forward until then because it's a language change and requires input from the working group.

**Comment by fiddlededee** [comment]
> In my opinion, the second column on the third row should have both the `amount` and the `sum` role.

Seems like most straightforward way.

**Comment by sindzicat** [comment]
Hello. What I need to do to get these new features?

**Comment by Mogztter** [comment]
@sindzicat I guess you could use my fork of Asciidoctor: https://github.com/asciidoctor/asciidoctor/pull/3539
Please note that my current implementation does not allow to:
- define anything other than roles `[.role1.role2]` (ie. you cannot define id)
- define roles on rows (only on cells and columns)

**Comment by brunchboy** [comment]
I’m so glad the spec is starting to move, because this is what is holding me back from porting my diagrams out of LaTeX, and some of the TeX packages my documents use are unmaintained and no longer work with this year’s TeX.

**Comment by sindzicat** [comment]
@brunchboy, could you give a little more information about your problem? It would be fine if you will create a new issue for it. I believe we may solve your problem with current state of Asciidoctor.

**Comment by brunchboy** [comment]
Thanks, @sindzicat, I am certain it is possible with an extension in any case. @mojavelinux started me down that direction last year, so I created https://github.com/Deep-Symmetry/asciidoctor-multirole_table and he was going to point me at to how to actually hook it into Asciidoctor, since I am not a Ruby developer. But he has (quite understandably) been much too busy with more important and widely-useful tasks, so I have been stuck since.

**Comment by brunchboy** [comment]
One of the documents I am trying to port from LaTeX to Asciidoctor is here: https://deepsymmetry.org/dysentery/Analysis.pdf

Figure 31 at the top of page 33 is a good example of the kind of diagram I need this sort of multi-role table to implement in Asciidoctor.

**Comment by sindzicat** [comment]
@brunchboy, here is my solution:

![image](https://user-images.githubusercontent.com/33900895/74928373-6c246a80-53ea-11ea-86b3-e06d36263e67.png)

<details>
<summary>Code (click to view)</summary>

```
= Creating table with colored cell
:nofooter:

++++
<style>
td.tableblock {
    margin: 0;
    padding: 0;
}
td.tableblock p {
    padding: 0.5rem;
}
td.tableblock .bg-yellow p {
    background-color: yellow;
}
td.tableblock .bg-green p {
    background-color: lime;
}
td.tableblock .bg-pink p {
    background-color: pink;
}
td.tableblock .bg-blue p {
    background-color: RGB(96,197,241);
}
</style>
++++

[cols="^1a,^4a,^1a,^4a,^1a,^2a,^1a,^1a,^1a"]
|===
|
[.bg-green]
11

|
[.bg-green]
872349ae

|
[.bg-yellow]
11

|
[.bg-yellow]
TxID

|
[.bg-pink]
10

|
[.bg-pink]
4002

|
[.bg-blue]
0f

|
[.bg-blue]
04

| 14!

|===
```
</details>

Some words about `cols` notation.  `^` means center text horizontally. Digits means the relative width of column. Full width of table is 1+4+1+4+1+2+1+1+1=16. So, when you write 1, this means 1/16 of table width, 4 &mdash; 4/16 of table width and so on. Letter `a` after digit means that cell content is written in Asciidoc format.

**Comment by brunchboy** [comment]
Interesting, I did not realize that much could be tweaked inside the existing implementation. Unfortunately it does lead to a situation where the markup heavily outweighs the actual information, making the source document very hard to read, which somewhat defeats the promise of Asciidoc. Still, this is worth experimenting further with. Thank you!

Do you think I will be able to achieve the control over cell boundaries (to show when cells extend on to later lines) and the diagonal lines to indicate arbitrary length as well?

**Comment by brunchboy** [comment]
To clarify, my main question is, can I supply a list of multiple roles to each cell, such as:

`[.bg-blue,.border-left,.border-top]`

I think that was the issue that was requiring my extension, now that I am casting my mind back. It would be impractical to have to define separate role names for every combination of possibilities needed for rendering my packet diagrams.

**Comment by Mogztter** [comment]
> I think that was the issue that was requiring my extension, now that I am casting my mind back. It would be impractical to have to define separate role names for every combination of possibilities needed for rendering my packet diagrams.

For this particular use case you could probably use a diagram library no? I've recently added [packetdiag](http://blockdiag.com/en/nwdiag/packetdiag-examples.html) in https://kroki.io/ (not yet released)

**Comment by brunchboy** [comment]
I looked at packetdiag and found it lacking compared to bytefield back when I initially chose the LaTeX route. I’m not willing to give up any of the design details of my diagrams.

**Comment by Mogztter** [comment]
I'm not saying that you should but I think a diagram library should be more "appropriated" than a table no?

> found it lacking compared to bytefield

Unfortunately, and as far as I know, `bytefield` is tightly coupled to LaTeX. Do you know if there's a good alternative?
Sorry I digress 😉

**Comment by brunchboy** [comment]
Yes, sadly it is. I thought about porting something like `bytefield` to the Asciidoctor ecosystem, but realized it would take far too much time away from building the actual features of my projects, and working in languages that are not my areas of comparative advantage.

I would love to be able to find something that could draw my diagrams using a concise specification domain-specific language rather than hacking them with tables and CSS, but that doesn’t exist today, and I can’t take the time to create it myself, nor do I have revenue that would allow me to pay someone else to create it, so that’s why I am stuck. 😄

**Comment by sindzicat** [comment]
@brunchboy, yes, you can get unlimited control over the table via CSS. I means you can really use something like `[.bg-blue.border-left.border-top]`. Please note that commas between CSS class names are not required. You even can use id property! For example `[#cell-11.bg-blue]`.

Also, you're definitely right about that hacked table became more difficult to read. But this table is still easier than a raw HTML. I have very complex case with application forms, and my `.adoc` became so difficult to read, that I just switched to HTML for these forms. The main thing for me is the _balance_ between effort and readability of the text.

I'd like to say that Asciidocs (Asciidoctor) is the most powerfull tool for technical documentation. But unfortunately, it cannot do anything we wish. It has own bounds as well. So I haven't see any problems to hack something using CSS or even use pure HTML to achieve the goal.

**Comment by brunchboy** [comment]
Thank you for the confirmation @sindzicat, I look forward to attempting this approach!

**Comment by Mogztter** [comment]
> I would love to be able to find something that could draw my diagrams using a concise specification domain-specific language rather than hacking them with tables and CSS, but that doesn’t exist today, and I can’t take the time to create it myself, nor do I have revenue that would allow me to pay someone else to create it, so that’s why I am stuck. 😄

I think GraphViz could do it, take a look at:
-  https://imgur.com/gallery/LWOGq
- https://renenyffenegger.ch/notes/tools/Graphviz/examples/index
- https://www.graphviz.org/doc/info/shapes.html

(digress again)

**Comment by brunchboy** [comment]
I looked at GraphViz. I don’t see anything in those examples that suggest it can handle the cells-with-open-borders needed to clearly indicate sections which don’t align with row boundaries, or the diagonal-lines-to-indicate-discontinuity, or the multi-row and multi-column header groupings. If any graphviz expert wants to prove me wrong by precisely reconstructing my diagrams using it and showing me how, I’d be happy to learn, but I am not willing to invest my own time chasing down what looks to me like a dead end.

**Comment by brunchboy** [comment]
Well, @sindzicat I was almost able to get your approach working, but there seems to be one fatal flaw. When I assign roles/styles to the individual cells, they get applied at the level of the `div` _inside_ the `td`, not to the `td` itself, so I cannot override the left, right, top, or bottom border, which is a critical per-cell requirement for me. I think that is why @mojavelinux said this requires an Asciidoctor extension.

**Comment by sindzicat** [comment]
@brunchboy, yes, extension is better and you control borders too. I may create example for this case too.

**Comment by sindzicat** [comment]
I got some solution, but it is very ugly. But I found [here](https://asciidoctor.org/docs/asciidoc-syntax-quick-reference/#more-delimited-blocks) the following:

![image](https://user-images.githubusercontent.com/33900895/75010815-2c638e80-548f-11ea-8d10-230cec11adba.png)

So I'll try solution on js.

**Comment by sindzicat** [comment]
Stop! Using js means, that content, generated Asciidoctor, isn't full. We need open generated HTML in browser, wait for js to be fully executed and only after that we will get content we need. Not good. Then there are two ways:

1. generate HTML via Asciidoctor, then do HTML post processing on any programming language you know,
2. write extension

I may post process HTML, but on Python only...

**Comment by brunchboy** [comment]
So I ended up trying to build my diagrams using just raw hand-created HTML and CSS, and again almost go them working, but ran into what appears to be a weird bug in the table rendering engine where cells that were configured not to have borders on their bottom side were getting borders when cells were added below that were configured not to have borders on their top side.

So I finally gave up (also, the HTML was really big and horrible to write, even if it had worked), and just decided to write a Clojure program that could generate the diagrams as SVG. And I have that working well enough to recreate the diagram I linked to above. But now I even have more control over it than I did in LaTeX, and have been able to make it more compact but as informative. So I am happy with this direction. I will either generate the SVG and check them into Git, or look into wrapping this program inside asciidoctor-diagram, so I can embed the Clojure source inside my Asciidoctor page and have the diagrams rendered in-situ. Which would be cool, and then let other people more easily make such diagrams as well.

I tried uploading the resulting SVG here, but GitHub doesn’t yet support that format. Here it is on my web server: https://deepsymmetry.org/images/test.svg

**Comment by Mogztter** [comment]
The end result looks really good 👍
Could you please share the syntaxe/grammar used to produce this diagram?

If you plan to open source your code I can integrate it in https://kroki.io/ 🚀

**Comment by danyill** [comment]
One use case I have for add roles/ids to cells is so that I can map through a line number. In the vs-code Asciidoc extension we do something like this to retrieve line numbers to allow scroll syncing:

```js
const blocksWithLineNumber = ascii_doc.findBy(function (b) { return typeof b.getLineNumber() !== 'undefined'; });

blocksWithLineNumber.forEach(function (block, key, myArray) {
    block.addRole("data-line-" + block.getLineNumber());
})
```

With tables, because `data-line` doesn't get converted for cells/rows, the whole table only gets allocated a line number only for the table block element which means the user experience is that editor and preview synch is lost as the editor "skips" over the entire table.

All of which is to say, I'd be delighted if #3539 got merged soon :wink:

**Comment by brunchboy** [comment]
I ended up responding to @Mogztter on the Asciidoctor Gitter channel. My code for an update of the LaTeX `bytefield` package is already open-source, but it’s at an early proof of concept stage and in a lot of flux right now as I come up with a DSL that will be safe to evaluate in the context of kroki.io. It is looking very promising, and an expected result is an Asciidoctor extension as well.

**Comment by enricodarodda** [comment]
Good evening to everyone,
I'm sorry to bother, but I just wanted to understand if there is a way to specify and render the colours of a table (cell background colour, contours color, ...) in PDF format, using asciidoctor-pdf to elaborate the asciidoc file.
Thanks in advance :)

**Comment by Hexdump74** [comment]
Hi,

I have a solution for adding a background color. It fits for me, as (in my opinion) it is readable, pretty simple and easy.

Lets create a table :

```
|====
| [.danger]#Lick the cables# | [.safe]#Lick the lollipop#
|====
```

Add the following css code:

```
td:has(span.danger) {
  background-color: red;
}
td:has(span.safe) {
  background-color: green;
}
```

Hope it helps

**Comment by goyalyashpal** [comment]
> td:has(span.danger) {
>   background-color: red;
> }

hey @Hexdump74 !  can this be done in place in the `.adoc` file itself?

**Comment by Hexdump74** [comment]
No, you'll have to put that in your css file. Or maybe you can use a passthrough block to pass css from the adoc ? Didn't try.
